src/snark0.ml

```diff
@@ -1066,6 +1066,311 @@ module Make_basic (Backend : Backend_intf.S) = struct
+
+      let run_unchecked ~run ?exposing ?handler proof_system s =
+        let {prover_state; _} = proof_system in
+        proof_system.prover_state
+        <- {prover_state with eval_constraints= false; system= None} ;
+        run_with_input ~run ?exposing ?handler proof_system s
+
+      let run_checked' ~run ?exposing ?handler proof_system s =
+        let {prover_state; _} = proof_system in
+        let system = R1CS_constraint_system.create () in
+        proof_system.prover_state
+        <- {prover_state with eval_constraints= true; system= Some system} ;
+        match run_with_input ~run ?exposing ?handler proof_system s with
+        | exception e -> Or_error.of_exn e
+        | s, x -> Ok (s, x)
+
+      let run_checked ~run ?exposing ?handler proof_system s =
+        Or_error.map (run_checked' ~run ?exposing ?handler proof_system s)
+          ~f:(fun (s, x) ->
+            let s', x =
+              As_prover.run x
+                (Checked.Runner.get_value proof_system.prover_state)
+                s
+            in
+            (s', x) )
+
+      let check ~run ?exposing ?handler proof_system s =
+        Or_error.is_ok (run_checked' ~run ?exposing ?handler proof_system s)
+
+      let read_proving_key proof_system =
+        match proof_system.proving_key_path with
+        | Some path -> Some (Proving_key.of_string (In_channel.read_all path))
+        | None -> None
+
+      let write_proving_key proof_system key =
+        match proof_system.proving_key_path with
+        | Some path ->
+            Out_channel.write_all path ~data:(Proving_key.to_string key)
+        | None -> ()
+
+      let read_verification_key proof_system =
+        match proof_system.verification_key_path with
+        | Some path ->
+            Some (Verification_key.of_string (In_channel.read_all path))
+        | None -> None
+
+      let write_verification_key proof_system key =
+        match proof_system.verification_key_path with
+        | Some path ->
+            Out_channel.write_all path ~data:(Verification_key.to_string key)
+        | None -> ()
+
+      let prove ~run ?exposing ?proving_key ?handler proof_system s =
+        let {prover_state; _} = proof_system in
+        if
+          Option.is_some exposing || Option.is_some handler
+          || (not !(proof_system.has_run_with_input))
+          || Option.is_none proof_system.prover_state.system
+        then (
+          ( match exposing with
+          | Some exposing -> set_inputs proof_system exposing
+          | None -> () ) ;
+          let system = R1CS_constraint_system.create () in
+          proof_system.prover_state
+          <- {prover_state with eval_constraints= false; system= Some system} ;
+          ignore (run_with_input ~run ?exposing ?handler proof_system s) ) ;
+        let {Checked.Runner.input; aux; _} = prover_state in
+        let proving_key =
+          match proving_key with
```

Comment:
Consider using something else for the control flow here to avoid repetition like
```ocaml
List.find_map ~f:Lazy.force
  [ lazy proving_key
  ; lazy proof_system.proving_key
  ; lazy (read_proving_key proof_system) ]
```